Block-list for peers based on gossipsub score

diegomrsantos opened the pull request:
## Proposed Changes

Introduces a peer block-list based on libp2p gossipsub scores with automatic unblocking after a configurable epoch and exposes manual block/unblock APIs.

- Integrates `allow_block_list` in `PeerManager` to deny connections and track block timestamps.
- Adds periodic score checks in `Network` to auto-block low-scoring peers and disconnect them.
- Updates `PeerManager::new` signature to accept epoch duration, adjusts behaviour initialization, and adds tests.

diegomrsantos committed:
block peers

diegomrsantos committed:
Merge branch 'unstable' into block-peers

diegomrsantos commented on anchor/network/src/peer_manager.rs:
```diff
@@ -405,6 +499,9 @@ impl NetworkBehaviour for PeerManager {
 
     fn on_swarm_event(&mut self, event: FromSwarm) {
+        // Handle block list events first
+        self.block_list.on_swarm_event(event);
```
`FromSwarm` implements `Copy`

diegomrsantos committed:
remove unnecessary functions

dknopik commented on anchor/network/src/network.rs:
```diff
@@ -258,6 +267,10 @@ impl<R: MessageReceiver> Network<R> {
                         }
                     }
                 }
+                _ = self.peer_score_check_timer.tick() => {
+                    // Periodic peer score checks
+                    self.check_and_block_peers_by_score();
+                }
```
It would be nice to have only one heartbeat. I understand that we need access to the swarm for `check_and_block_peers_by_score`. Wdyt about moving the peer store heartbeat here also?

diegomrsantos replied:
Great idea. I changed it, let me know what you think.

dknopik reviewed (commented):
Neat, I like it!

diegomrsantos committed:
add tests

diegomrsantos committed:
remove unused libp2p-allow-block-list

diegomrsantos committed:
Refactor peer score checking mechanism

- Remove periodic peer score check timer from Network struct
- Add PeerManagerHeartbeat event containing connect actions and check_peer_scores flag
- Update PeerManager to emit heartbeat event with both connect actions and peer score check signal
- Update Network event handling to process new PeerManagerHeartbeat event
- Clean up unused imports in network.rs

This consolidates peer management actions into a single heartbeat event from PeerManager,
eliminating the separate periodic timer for peer score checks.

diegomrsantos committed:
improve naming

diegomrsantos commented:
Closing in favor of https://github.com/sigp/anchor/pull/432
